fl4health/parameter_exchange/layer_exchanger.py

```diff
@@ -92,88 +91,76 @@ def pull_parameters(self, parameters: NDArrays, model: nn.Module, config: Option
+    ) -> None:
+        self.layer_selection_function = layer_selection_function
 
     def set_filter_mode(self, filter_by_percentage: bool) -> None:
         self.filter_by_percentage = filter_by_percentage
 
     def set_normalization_mode(self, normalize: bool) -> None:
         self.normalize = normalize
 
+    def set_select_drift_more(self, select_drift_more: bool) -> None:
+        self.select_drift_more = select_drift_more
+
     def _calculate_drift_norm(self, t1: torch.Tensor, t2: torch.Tensor) -> float:
         t_diff = (t1 - t2).float()
         drift_norm = torch.linalg.norm(t_diff)
         if self.normalize:
             drift_norm /= torch.numel(t_diff)
         return drift_norm.item()
 
-    def filter_layers_by_threshold(self, model: nn.Module, initial_model: nn.Module) -> Tuple[NDArrays, List[str]]:
-        """
-        Return those layers of model that deviate (in l2 norm) away from corresponding layers of
-        self.initial_model by at least self.threshold.
-        """
-        layer_names = []
-        layers_to_transfer = []
-        initial_model_states = initial_model.state_dict()
-        model_states = model.state_dict()
-        for layer_name, layer_param in model_states.items():
-            layer_param_past = initial_model_states[layer_name]
-            drift_norm = self._calculate_drift_norm(layer_param, layer_param_past)
-            if drift_norm >= self.threshold:
-                layers_to_transfer.append(layer_param.cpu().numpy())
-                layer_names.append(layer_name)
-        return layers_to_transfer, layer_names
-
     def set_threshold(self, new_threshold: float) -> None:
-        self.threshold = new_threshold
+        self.norm_threshold = new_threshold
 
     def set_percentage(self, new_percentage: float) -> None:
         assert 0 < new_percentage <= 1
         self.exchange_percentage = new_percentage
 
-    def filter_layers_by_percentage(self, model: nn.Module, initial_model: nn.Module) -> Tuple[NDArrays, List[str]]:
-        names_to_norm_drift = {}
-        initial_model_states = initial_model.state_dict()
-        model_states = model.state_dict()
-
-        for layer_name, layer_param in model_states.items():
-            layer_param_past = initial_model_states[layer_name]
-            drift_norm = self._calculate_drift_norm(layer_param, layer_param_past)
-            names_to_norm_drift[layer_name] = drift_norm
-
-        total_param_num = len(names_to_norm_drift.keys())
-        num_param_exchange = int(math.ceil(total_param_num * self.exchange_percentage))
-        param_to_exchange_names = sorted(names_to_norm_drift.keys(), key=lambda x: names_to_norm_drift[x])[
-            :(num_param_exchange)
-        ]
-        return [model_states[name].cpu().numpy() for name in param_to_exchange_names], param_to_exchange_names
+    def select_parameters(
+        self, model: nn.Module, initial_model: Optional[nn.Module] = None
```

Comment:
For the signature here, I think you want to _force_. initial_model to have type `nn.Module` in the typing (ie. not optional) as the assertion below implies that it is required. In push_parameters, where this function is called, you already do an assertion, so the typing here can be stronger.